# <mark style="display:block; background:#d1c4e9; color:#1a1a1a; padding:6px 12px; border-radius:4px">5주차 · 종목을 먼저 추려 놓고 그 안에서만 순위를 매기기</mark>

지난주에는 한 사람에게 추천을 만들려고 **종목 100개를 전부** 점수 매겼습니다.

오늘은 순서를 바꿉니다. **먼저 계산이 적게 드는 방법으로 40개만 추리고(후보 생성), 그 40개만 지난주 모델로 정렬합니다(순위 매기기).**

**모델은 한 글자도 고치지 않습니다.** 바뀌는 것은 **점수를 매길 종목 개수** 하나뿐입니다.

---

### 오늘의 구성

| 파트 | 어디서 | 하는 일 |
|---|---|---|
| 개념 | 슬라이드 | 2단계 추천 · 후보 Recall 이라는 상한선 · Two-Tower · ANN |
| 실습 | **이 노트북** | 후보를 뽑고, 후보에만 점수를 매기고, 후보 개수를 바꿔 본다 |
| 마무리 | 슬라이드 | 오늘의 용어 · 점수판 · 다음 주 |

## <mark style="display:block; background:#c8e6c9; color:#1a1a1a; padding:6px 12px; border-radius:4px">시작하기 · 이 노트북을 여는 법</mark>

1. https://colab.research.google.com/github/welovecherry/recsys/blob/main/notebooks/05_two_stage.ipynb
2. 구글 계정으로 로그인합니다.
3. **경고창이 뜨면 `Run anyway` 를 누릅니다.**
4. **아래 셀의 ▶ 버튼을 누릅니다.** 실습 자료를 받아옵니다. 10초쯤 걸립니다.
5. 그다음부터는 위에서 아래로 셀을 하나씩 실행하면 됩니다.

> ⚠ **고친 내용을 남기려면** 메뉴에서 `파일 → 드라이브에 사본 저장` 을 눌러 주세요.

---

**이 노트북이 쓰는 것**

| 무엇 | 쓰임 |
|---|---|
| `data/items.csv` | 종목 100개 — 이름 · 섹터 · 위험도 |
| `data/users.csv` | 투자자 300명 |
| `data/interactions.csv` | 거래 기록 8,086건 |
| `recsys.py` | 데이터 읽기 · 시간순 분할 · 채점 · 2주차 세그먼트 |

⚠ **지난주 노트북을 먼저 돌릴 필요가 없습니다.** 이 노트북 하나로 끝납니다.

---

## <mark style="display:block; background:#c8e6c9; color:#1a1a1a; padding:6px 12px; border-radius:4px">먼저 이 셀부터 실행하세요  ▶</mark>

In [1]:
# 이 셀에서 하는 일 — Colab 에서 열었으면 실습 자료를 내려받는다
# 왜 하나 — Colab 은 열 때마다 빈 컴퓨터라 데이터·recsys.py 가 없다. 내 컴퓨터면 그냥 넘어간다

import os          # 폴더를 만들고 옮겨 다니는 도구
import sys         # 지금 파이썬이 어떤 환경인지 알려 주는 도구
import subprocess  # 터미널 명령을 파이썬에서 대신 실행해 주는 도구

if "google.colab" in sys.modules:                    # Colab 이면 이 안이 실행된다
    if os.path.exists("/content/recsys"):            # 전에 받아 둔 것이 있으면 최신으로
        subprocess.run(["git", "-C", "/content/recsys", "pull", "-q", "--ff-only"])
    else:                                            # 처음이면 통째로 내려받는다
        subprocess.run(["git", "clone", "-q",
                        "https://github.com/welovecherry/recsys.git", "/content/recsys"])
    os.chdir("/content/recsys/notebooks")            # 노트북 폴더 안으로 이동
    print("준비 끝 —", os.getcwd(), "· 아래 셀부터 차례로 실행하세요.")
else:
    print("내 컴퓨터에서 실행 중입니다 —", os.getcwd(), "· 따로 받아올 것이 없습니다.")

내 컴퓨터에서 실행 중입니다 — /Users/hong/workspaces/org_physical-spark/course-recsys/notebooks · 따로 받아올 것이 없습니다.


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">오늘 나오는 말 네 개  `[PPT]`</mark>

1. **2단계 추천 (two-stage)** — 추천을 **후보 생성**과 **순위 매기기** 두 단계로 나누는 구조
2. **후보 생성 (candidate generation)** — 1단계. 계산이 적게 드는 방법으로 종목 100개를 **40개로** 줄인다
3. **순위 매기기 (ranking)** — 2단계. 추려진 **40개만** 지난주 모델로 정렬한다
4. **후보 Recall (candidate recall)** — 정답이 **후보 안에 남아 있는 비율**. 최종 점수가 넘을 수 없는 **상한선**

---

### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">준비 · 데이터를 읽고 지난주와 같은 기준으로 나눈다</mark>

**여기서 하는 일** — 데이터를 읽고, 3주차부터 쓰던 **시간순 분할**로 학습 구간과 채점 구간을 나눕니다.

**채점 함수는 다시 만들지 않습니다.** `recsys.recall_at_k` 가 이미 있습니다 — 3주차 방식 그대로입니다.
**채점 기준이 같아야** 오늘 숫자를 지난주 **0.3918** 과 비교할 수 있습니다.

In [2]:
# 이 셀에서 하는 일 — 도구를 부르고, 데이터를 읽고, 시간순으로 나눈다
# 왜 하나 — 지난주와 같은 기준으로 채점해야 0.3918 과 비교할 수 있다

import sys                                  # 파이썬이 파일을 찾는 경로를 다루는 도구
sys.path.insert(0, ".")                     # 지금 폴더에서 recsys.py 를 찾게 한다
sys.path.insert(0, "notebooks")             # 한 칸 안쪽 폴더도 찾게 한다

import numpy as np                          # 숫자 묶음을 빠르게 다루는 도구. np 라고 부른다
import recsys                               # 이 수업용 도구 모음 — 읽기·분할·채점이 들어 있다

items, users, interactions = recsys.load()                  # 종목·투자자·거래 기록 세 표
train, test, 기준시점 = recsys.split_by_time(interactions)   # 3주차부터 쓰던 시간순 분할

이름_사전 = items.set_index("item_id")["name"].to_dict()      # {종목 번호: 이름}
이미_담은것 = train.groupby("user_id")["item_id"].apply(set).to_dict()   # groupby = 사람별로 묶는다
전체_인기순위 = list(train["item_id"].value_counts().index)   # value_counts = 같은 값이 몇 번인지 센다

print(f"학습 {len(train):,}건 · 채점 {len(test):,}건 · 자른 날짜 {기준시점.date()}")

학습 7,095건 · 채점 991건 · 자른 날짜 2026-07-30


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.1 지난주 모델을 그대로 되살린다</mark>

**5.1 에서 하는 일** — 4주차에 만든 **사람표(285×8)** 와 **종목표(8×100)** 를 다시 만듭니다.

**1. 왜 다시 만드나**
- 노트북을 새로 열면 지난주 변수가 남아 있지 않습니다. 지난주 결과 파일을 읽어 오는 것이 아니라 **원본 데이터에서 다시 만듭니다.**
- 표를 만드는 한 줄은 `recsys.interaction_matrix` 로 내렸습니다. **4주차에 이미 손으로 만들어 봤기 때문입니다.**

**2. 오늘 모델은 여기서 끝입니다**
- 이 셀 뒤로는 **모델을 한 글자도 고치지 않습니다.**

**3. 숫자 읽는 법**
- **사람표** — 세로 사람 285명 × 가로 **잠재 요인 8개**
- **종목표** — 세로 **잠재 요인 8개** × 가로 종목 100개 (사람표와 방향이 반대)

In [3]:
# 이 셀에서 하는 일 — 4주차와 똑같이 표를 만들고 잠재 요인 8개로 쪼갠다
# 왜 하나 — 오늘 쓰는 모델이 지난주 것과 같다는 것을 확인하려는 것이다

from sklearn.decomposition import TruncatedSVD      # 4주차에 쓴 행렬분해 도구

표 = recsys.interaction_matrix(train, items)        # 사람 × 종목 표 — 담았으면 1, 아니면 0
모델 = TruncatedSVD(n_components=8, random_state=42)   # 잠재 요인 8개 — 지난주와 같은 설정
사람표 = 모델.fit_transform(표)                      # 285 × 8
종목표 = 모델.components_                            # 8 × 100

사람_이름들 = list(표.index)                         # 표의 세로 이름(사람) 목록
열이름 = list(표.columns)                            # 표의 가로 이름(종목 번호) 목록

print(f"사람표 {사람표.shape} · 종목표 {종목표.shape}   ← 지난주와 같은 크기")

사람표 (285, 8) · 종목표 (8, 100)   ← 지난주와 같은 크기


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.2 2주차 세그먼트 순위표를 가져온다</mark>

**5.2 에서 하는 일** — 2주차에 만든 **세그먼트**와 세그먼트별 인기 순위표를 가져옵니다.

**1. 세그먼트가 무엇이었나 (2주차)**
- 사람마다 **주력 섹터**(가장 많이 담은 섹터)와 **평균 위험도**로 이름을 붙인 것입니다 — `레버리지|4` 처럼.
- 285명이 **15개** 세그먼트로 나뉘었고, 그때 점수가 **0.2684** 였습니다.

**2. 왜 코드를 안 쓰고 한 줄로 부르나**
- 만드는 방법은 **2주차에 직접 타이핑해 봤습니다.** 오늘 배우는 것이 아니므로 `recsys.segments` 로 내렸습니다.
- **파일을 읽어 오는 것이 아닙니다.** 부르면 그 자리에서 원본 거래 기록으로 다시 만듭니다.

**3. 출력에서 확인할 것**
- 세그먼트마다 종목 수가 100개보다 적습니다. **그 세그먼트 사람이 한 번도 안 담은 종목**이 빠져 있기 때문입니다.
- 이 「빠짐」이 곧 1단계가 하는 걸러내기입니다.

In [4]:
# 이 셀에서 하는 일 — 2주차 세그먼트와 세그먼트별 인기 순위표를 만든다
# 왜 하나 — 오늘 후보를 뽑을 재료다. 2주차에 배운 것이라 한 줄로 부른다

세그먼트, 순위표 = recsys.segments(train, items)    # ({사람: 이름}, {이름: 종목 순서})

print(f"세그먼트 {len(순위표)}개 · U0003 은 {세그먼트['U0003']}\n")
for 이름 in ["레버리지|4", "지수|3", "채권|3"]:      # 큰 것과 작은 것을 하나씩 본다
    목록 = 순위표[이름]                              # 그 세그먼트의 종목 순서
    print(f"  {이름:8} 종목 {len(목록):3}개 · 1위 {이름_사전[목록[0]]}")

세그먼트 15개 · U0003 은 레버리지|4

  레버리지|4   종목  93개 · 1위 Direxion Semiconductor Bull 3X
  지수|3     종목 100개 · 1위 iShares Core MSCI EAFE ETF
  채권|3     종목  23개 · 1위 iShares MSCI USA Min Vol Factor ETF


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.3 후보 40개를 꺼낸다  ✏️ 직접 해 보기</mark>

**5.3 에서 하는 일** — 그 사람 세그먼트의 순위표에서 **위에서 40개**를 꺼냅니다. 여기가 **1단계**입니다.

**1. 여기서 모델을 쓰지 않습니다**
- 순위표는 이미 만들어져 있습니다. **꺼내 오기만 합니다 — 곱셈을 한 번도 안 한다.**

**2. 40개가 안 되는 세그먼트는**
- 종목이 20개뿐인 세그먼트도 있습니다. 그럴 때는 **전체 인기순으로 빈자리를 채웁니다**(`recsys.fill_up`).

**3. 직접 채울 부분**
- **위에서 몇 개를 자를지** 한 줄입니다. `[:40]` 은 **앞에서 40개까지**라는 뜻입니다.

In [5]:
# 이 셀에서 하는 일 — ✏️ 빈칸 1 · 순위표에서 후보 40개를 꺼낸다
# 왜 하나 — 여기가 1단계(후보 생성)다. 모델을 한 번도 쓰지 않는다

후보_개수 = 40                                       # 5.7 에서 이 숫자를 바꿔 본다

def 후보뽑기(사람, 이미):                              # def = 함수를 만든다
    """그 사람 세그먼트의 순위표에서 위에서 40개를 돌려준다."""
    목록 = 순위표.get(세그먼트.get(사람), [])          # get = 없으면 기본값을 준다
    후보 = 목록[:후보_개수]                            # ← ✏️ 빈칸 : 위에서 몇 개를 자를까
    return recsys.fill_up(후보, 전체_인기순위, 후보_개수)   # 모자라면 전체 인기순으로 채운다


내_후보 = 후보뽑기("U0003", 이미_담은것["U0003"])        # 만들었으면 한 명 돌려 본다
print(f"U0003 의 후보 {len(내_후보)}개 중 앞 3개")
for 번호 in 내_후보[:3]:                               # [:3] = 앞에서 3개만
    print("  ", 이름_사전[번호])

U0003 의 후보 40개 중 앞 3개
   Direxion Semiconductor Bull 3X
   TIGER 미국나스닥100레버리지
   KODEX 코스닥150레버리지


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.4 후보 Recall — 오늘의 핵심 숫자</mark>

**5.4 에서 하는 일** — 후보 40개 안에 **정답이 몇 개 남아 있는지** 잽니다.

**1. 후보 Recall 이 무엇인가**
- 어떤 분이 다음 달에 종목 4개를 담았는데, 후보 40개 안에 그중 2개만 있으면 **후보 Recall 은 0.5** 입니다.

**2. 왜 이것이 상한선인가**
- 후보에 없는 2개는 **점수조차 매기지 않습니다.** 추천 목록에 올라갈 방법이 없습니다.
- 그래서 **최종 Recall@10 은 후보 Recall 을 절대 넘지 못합니다.**
- **1단계가 놓친 것은 2단계가 살려낼 수 없습니다.** 오늘 가장 중요한 한 줄입니다.

**3. 어떻게 재나**
- 채점 함수를 새로 만들지 않습니다. `recsys.recall_at_k` 에 **추천 함수 대신 후보 함수**를 넣고 `k` 를 40으로 줍니다.
- `recsys.recall_at_k` 는 `recsys.py` 파일에 **미리 만들어 둔 채점 도구**입니다. 반면 바로 위 5.3 에서 만든 `후보뽑기` 는 **오늘 이 노트북에서 직접 짠 함수**입니다.

In [6]:
# 이 셀에서 하는 일 — 후보 40개 안에 정답이 얼마나 남았는지 잰다
# 왜 하나 — 이 값이 오늘 점수의 상한선이다. 여기보다 높은 점수는 나올 수 없다

후보_recall, 인원 = recsys.recall_at_k(후보뽑기, train, test, k=후보_개수)

print(f"후보 {후보_개수}개의 후보 Recall = {후보_recall:.4f}   (채점 대상 {인원}명)")
print(f"→ 정답의 {후보_recall:.1%} 만 후보에 있습니다.")            # .1% = 소수 첫째 자리 퍼센트
print(f"→ 남은 {1 - 후보_recall:.1%} 는 점수조차 안 매깁니다. 오늘 못 맞힙니다.")

후보 40개의 후보 Recall = 0.5287   (채점 대상 266명)
→ 정답의 52.9% 만 후보에 있습니다.
→ 남은 47.1% 는 점수조차 안 매깁니다. 오늘 못 맞힙니다.


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.5 후보 40개에만 점수를 매긴다  ✏️ 직접 해 보기</mark>

**5.5 에서 하는 일** — 후보 40개**만** 지난주 모델로 점수 매겨 순위를 정합니다. 여기가 **2단계**입니다.

**지난주 코드와 딱 한 군데 다릅니다.**

| | 곱하는 줄 | 점수를 매기는 개수 |
|---|---|---|
| 4주차 | `사람표[내자리] @ 종목표` | 100개 |
| **5주차** | `사람표[내자리] @ 종목표[:, 후보자리]` | **40개** |

- `[:, 후보자리]` 는 **「세로는 전부, 가로는 후보 자리만」** 이라는 뜻입니다. 결과가 `8 × 40` 이 됩니다.
- **직접 채울 부분** — 후보 종목이 **표에서 몇 번째 가로 칸**인지 찾는 줄입니다.

In [7]:
# 이 셀에서 하는 일 — ✏️ 빈칸 2 · 후보 40개에만 점수를 매겨 위에서 10개를 고른다
# 왜 하나 — 여기가 2단계(순위 매기기)다. 모델은 지난주 것 그대로고 개수만 줄었다

def 추천하기(사람, 이미):
    """후보 40개에만 점수를 매겨 높은 순으로 10개를 돌려준다."""
    if 사람 not in 사람_이름들:                        # 표에 줄이 없는 사람(신규)은
        return recsys.take(전체_인기순위, 이미)[:10]    # 전체 인기 목록으로 준다

    후보 = 후보뽑기(사람, 이미)                         # 1단계 — 40개로 줄인다
    후보자리 = [열이름.index(번호) for 번호 in 후보]     # ← ✏️ 빈칸 : index = 몇 번째인지 찾는다
    점수 = 사람표[사람_이름들.index(사람)] @ 종목표[:, 후보자리]   # 세로 전부, 가로는 후보만

    내_순서 = [후보[j] for j in np.argsort(-점수)]      # argsort(-값) = 큰 것부터 자리 번호
    return recsys.take(내_순서, 이미)[:10]              # 이미 담은 것을 빼고 위에서 10개


print("U0003 에게 줄 추천 3개")
for 번호 in 추천하기("U0003", 이미_담은것["U0003"])[:3]:
    print("  ", 이름_사전[번호])

U0003 에게 줄 추천 3개
   TIGER 미국나스닥100레버리지
   Salesforce
   iShares Global Clean Energy ETF


### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.6 지난주와 같은 기준으로 채점한다</mark>

**5.6 에서 하는 일** — 방금 만든 2단계 추천을 3주차와 같은 `Recall@10` 으로 잽니다.

**미리 말씀드립니다 — 점수가 떨어집니다.**
- 「계산을 줄였으니 점수가 좀 떨어지겠지」가 맞습니다. **잘못 한 것이 아닙니다.**
- 왜 떨어지는지는 다음 칸의 표에서 설명합니다.

In [8]:
# 이 셀에서 하는 일 — 3주차와 같은 기준으로 채점한다 (한 줄)
# 왜 하나 — 같은 함수를 써야 지난주 0.3918 과 공정하게 비교할 수 있다

이단계_점수, 인원 = recsys.recall_at_k(추천하기, train, test)   # 3주차 방식 그대로

print(f"5주차 2단계 추천 — Recall@10 = {이단계_점수:.4f}   (채점 대상 {인원}명)")
print(f"4주차 전부 계산   — Recall@10 = 0.3918")                # 견줄 상대
print(f"\n→ {이단계_점수 - 0.3918:+.4f} · 점수는 떨어지고, 계산은 100개에서 {후보_개수}개로 줄었습니다.")

5주차 2단계 추천 — Recall@10 = 0.3382   (채점 대상 266명)
4주차 전부 계산   — Recall@10 = 0.3918

→ -0.0536 · 점수는 떨어지고, 계산은 100개에서 40개로 줄었습니다.


**결과 — 계산을 줄인 만큼 점수를 잃었습니다**

| 주차 | 방법 | 계산한 개수 | Recall@10 |
|---|---|---|---|
| 2·3주차 | 세그먼트별 인기 목록 | — | 0.2684 |
| 4주차 | 행렬분해 · 종목 전부에 점수 | 100개 | **0.3918** |
| **5주차** | **2단계 추천** · 후보 생성 + 순위 매기기 | **40개** | **0.3382** |

**점수가 떨어진 것이 정상입니다. 이번 주가 얻는 것은 점수가 아닙니다.**

- 4주차 모델은 종목 100개를 **전부** 보고 가장 좋은 10개를 골랐습니다. 고를 수 있는 폭이 가장 넓었습니다.
- 오늘은 그 모델에게 **40개만** 보여줬습니다. 나머지 60개는 **점수조차 매기지 않으니** 그 안에 정답이 있어도 추천 목록에 못 올라갑니다.

**그러면 왜 이렇게 하나요.**
- 종목이 100개일 때는 이 거래가 **손해**입니다. 전부 계산해도 21밀리초니까요.
- 종목이 30만 개가 되면 **전부 계산하는 쪽이 불가능**해집니다. 그때는 선택지가 이것뿐입니다.

### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.7 후보 개수를 바꿔 본다</mark>

**5.7 에서 하는 일** — 후보를 10·40·100개로 바꿔 세 번 돌려 봅니다.

**1. 첫 줄이 이 주차의 핵심입니다**
- 후보 10개일 때 상한선이 0.1242 인데 실제 점수도 **0.1242** 입니다. **상한선을 100% 썼습니다.**
- **2단계는 할 수 있는 일을 완벽하게 했습니다.** 그런데도 점수는 가장 낮습니다.
- 점수가 낮은 이유가 **순위를 잘못 매겨서가 아니라 후보에 정답이 없어서**라는 뜻입니다.

**2. 마지막 줄은 코드 검증 장치입니다**
- 후보를 **100개**로 잡으면 종목 전체와 같습니다. 즉 **1단계가 아무것도 걸러내지 않습니다.**
- 따라서 **0.3918 이 나와야 맞습니다.** **다른 숫자가 나오면 5.3이나 5.5를 잘못 채운 것입니다.**

In [9]:
# 이 셀에서 하는 일 — 후보 개수를 바꿔 가며 상한선과 최종 점수를 함께 잰다
# 왜 하나 — 점수가 낮은 이유가 1단계인지 2단계인지 가려내려는 것이다

print(f"{'후보':>4} {'상한선':>8} {'Recall@10':>10} {'상한선을 얼마나':>14}")
for 개수 in [10, 40, 100]:
    후보_개수 = 개수                                    # 위 두 함수가 이 값을 보고 움직인다
    상한선, _ = recsys.recall_at_k(후보뽑기, train, test, k=개수)
    점수, _ = recsys.recall_at_k(추천하기, train, test)
    print(f"{개수:>4} {상한선:>8.4f} {점수:>10.4f} {점수/상한선:>13.0%}")

후보_개수 = 40                                          # 실험이 끝났으니 원래 값으로 되돌린다

  후보      상한선  Recall@10       상한선을 얼마나
  10   0.1242     0.1242          100%
  40   0.5287     0.3382           64%


 100   1.0000     0.3918           39%

**출력 읽는 법**

1. **후보 10개** — 상한선 0.1242, 점수도 0.1242. **상한선을 100% 썼는데 점수는 가장 낮습니다.**
   - **1단계가 놓친 것은 2단계가 살려낼 수 없습니다.**
2. **후보 100개** — 상한선 1.0000, 점수 **0.3918**. 지난주와 같습니다.
   - **여기가 0.3918 이면 오늘 코드를 제대로 짠 것입니다.**

**한 줄로 정리하면** — 후보를 줄이면 **점수를 잃고 계산을 법니다.**

### <mark style="display:block; background:#f8bbd0; color:#1a1a1a; padding:6px 12px; border-radius:4px">5.8 오늘의 마무리  `[PPT]`</mark>

1. 2주차 **세그먼트 15개**의 순위표에서 후보 40개를 꺼내 **1단계(후보 생성)** 를 했습니다. 곱셈을 한 번도 안 한다입니다.
2. **후보 Recall = 0.5287** 이라는 상한선을 확인했습니다.
3. 그 40개에만 지난주 모델을 돌려 **Recall@10 = 0.3382** 를 얻었습니다.
4. 후보 10개일 때 **상한선을 100% 쓰고도 점수가 가장 낮은 것**을 확인했습니다.

**오늘 남는 한 줄** — **1단계가 놓친 것은 2단계가 살려낼 수 없습니다.**

다음 주에는 지금까지 모델에 한 번도 넣지 않은 **종목의 섹터·자산군·위험도·테마**를 섞습니다.

In [10]:
# 이 셀에서 하는 일 — 오늘 점수를 점수판에 남긴다
# 왜 하나 — 점수가 떨어진 주차도 그대로 적는다. 왜 떨어졌는지 설명할 수 있으면 된다

recsys.record(5, "2단계 추천 · 후보 생성 + 순위 매기기", 이단계_점수,
              note="2주차 세그먼트로 후보 40개를 추리고 그 40개만 4주차 모델로 정렬했다")

recsys.leaderboard(upto=5)   # 오늘까지 쌓인 점수판 (뒤 주차는 빼고 본다)

레벨 5 · 2단계 추천 · 후보 생성 + 순위 매기기 · Recall@10 = 0.3382


,level,name,recall_at_10,note
0,1,모두에게 같은 인기 순위,0.2163,알고리즘 없음. 인기 상위 10개를 모두에게 같게 추천했다
1,2,취향이 비슷한 세그먼트끼리,0.2684,거래 기록으로 주력 섹터와 평균 위험도를 뽑아 세그먼트로 나눴다
2,3,세그먼트별 목록 — 2주차와 같음,0.2684,추천 방식은 2주차와 같다. 무작위로 나누면 0.3150 이 나오는데 그것은 미래를...
3,4,행렬분해 SVD · 요인 8개,0.3918,규칙을 사람이 정하지 않았다. 사람 285 × 종목 100 표를 요인 8개로 쪼갰다
4,5,2단계 추천 · 후보 생성 + 순위 매기기,0.3382,2주차 세그먼트로 후보 40개를 추리고 그 40개만 4주차 모델로 정렬했다
